src/ucs/datastruct/ptr_map.inl

```diff
@@ -95,71 +121,166 @@ ucs_ptr_map_put(ucs_ptr_map_t *map, void *ptr, int indirect,
         return UCS_ERR_NO_PROGRESS;
     }
 
+    if (is_put_thread_safe) {
+        hash = &safe_hash->hash;
+    } else {
+        hash = &map->hash;
+    }
+
+    if (is_put_thread_safe) {
+        ucs_spin_lock(&safe_hash->lock);
+    }
+
     *key = (map->next_id += UCS_PTR_MAP_KEY_MIN_ALIGN) |
            UCS_PTR_MAP_KEY_INDIRECT_FLAG;
+    status = ucs_ptr_hash_put(hash, *key, ptr);
 
-    iter = kh_put(ucs_ptr_map_impl, &map->hash, *key, &ret);
-    if (ucs_unlikely(ret == UCS_KH_PUT_FAILED)) {
-        return UCS_ERR_NO_MEMORY;
-    } else if (ucs_unlikely(ret == UCS_KH_PUT_KEY_PRESENT)) {
-        return UCS_ERR_ALREADY_EXISTS;
+    if (is_put_thread_safe) {
+        ucs_spin_unlock(&safe_hash->lock);
     }
 
-    kh_value(&map->hash, iter) = ptr;
-    return UCS_OK;
+    return status;
 }
 
-/**
- * Get a pointer value from the map by its key.
- *
- * @param [in]  map     Container to get the pointer value from.
- * @param [in]  key     Key to look up in the container.
- * @param [in]  extract Whether to remove the key from the map.
- * @param [out] ptr_p   If successful, set to the pointer found in the map.
- *
- * @return UCS_OK if found, UCS_ERR_NO_PROGRESS if this key is direct and
- *         therefore no action was performed, UCS_ERR_NO_ELEM if not found.
- */
 static UCS_F_ALWAYS_INLINE ucs_status_t
 ucs_ptr_map_get(ucs_ptr_map_t *map, ucs_ptr_map_key_t key, int extract,
-                void **ptr_p)
+                void **ptr_p, int is_put_thread_safe,
+                ucs_ptr_safe_hash_t *safe_hash)
 {
-    khiter_t iter;
+    ucs_status_t status;
 
     if (ucs_likely(!ucs_ptr_map_key_indirect(key))) {
         *ptr_p = (void*)key;
         return UCS_ERR_NO_PROGRESS;
     }
 
-    iter = kh_get(ucs_ptr_map_impl, &map->hash, key);
-    if (ucs_unlikely(iter == kh_end(&map->hash))) {
-        return UCS_ERR_NO_ELEM;
+    status = ucs_ptr_hash_get(&map->hash, key, extract, ptr_p);
+    if (ucs_likely(status != UCS_ERR_NO_ELEM)) {
+        return status;
     }
 
-    *ptr_p = kh_value(&map->hash, iter);
-    if (extract) {
-        kh_del(ucs_ptr_map_impl, &map->hash, iter);
+    if (is_put_thread_safe) {
```

Comment:
maybe put this part in a separate non-inline function to reduce code size at target?
(can add ptr_map.c file)

Reply:
Should we do the same for init, destroy and put?

Reply:
yes

Reply:
done